pkg/kubelet/pod_workers.go

```diff
@@ -1030,240 +1294,254 @@ func (p *podWorkers) acknowledgeTerminating(pod *v1.Pod) PodStatusFunc {
 	}
-}
 
-// contextForWorker returns or initializes the appropriate context for a known
-// worker. If the current context is expired, it is reset. If no worker is
-// present, no context is returned.
-func (p *podWorkers) contextForWorker(uid types.UID) context.Context {
+	// if there is a pending update for this worker, requeue immediately, otherwise
+	// clear working status
 	p.podLock.Lock()
 	defer p.podLock.Unlock()
-
-	status, ok := p.podSyncStatuses[uid]
-	if !ok {
-		return nil
-	}
-	if status.ctx == nil || status.ctx.Err() == context.Canceled {
-		status.ctx, status.cancelFn = context.WithCancel(context.Background())
+	if status, ok := p.podSyncStatuses[podUID]; ok {
+		if status.pendingUpdate != nil {
+			select {
+			case p.podUpdates[podUID] <- struct{}{}:
+				klog.V(4).InfoS("Requeueing pod due to pending update", "podUID", podUID)
+			default:
+				klog.V(4).InfoS("Pending update already queued", "podUID", podUID)
+			}
+		} else {
+			status.working = false
+		}
 	}
-	return status.ctx
 }
 
 // SyncKnownPods will purge any fully terminated pods that are not in the desiredPods
 // list, which means SyncKnownPods must be called in a threadsafe manner from calls
-// to UpdatePods for new pods. It returns a map of known workers that are not finished
-// with a value of SyncPodTerminated, SyncPodKill, or SyncPodSync depending on whether
-// the pod is terminated, terminating, or syncing.
-func (p *podWorkers) SyncKnownPods(desiredPods []*v1.Pod) map[types.UID]PodWorkerState {
+// to UpdatePods for new pods. Because the podworker is dependent on UpdatePod being
+// invoked to drive a pod's state machine, if a pod is missing in the desired list the
+// pod worker must be responsible for delivering that update. The method returns a map
+// of known workers that are not finished with a value of SyncPodTerminated,
+// SyncPodKill, or SyncPodSync depending on whether the pod is terminated, terminating,
+// or syncing.
+func (p *podWorkers) SyncKnownPods(desiredPods []*v1.Pod) (map[types.UID]PodWorkerState, map[types.UID]struct{}) {
 	workers := make(map[types.UID]PodWorkerState)
 	known := make(map[types.UID]struct{})
 	for _, pod := range desiredPods {
 		known[pod.UID] = struct{}{}
 	}
+	var restartable map[types.UID]struct{}
 
 	p.podLock.Lock()
 	defer p.podLock.Unlock()
 
 	p.podsSynced = true
 	for uid, status := range p.podSyncStatuses {
-		if _, exists := known[uid]; !exists || status.restartRequested {
-			p.removeTerminatedWorker(uid)
+		// we only remove workers that are not represented in our config in order to
+		// prevent accidental restarts of previously terminated pods
+		_, knownPod := known[uid]
+		orphan := !knownPod
+
+		switch {
+		case status.restartRequested:
+			if restartable == nil {
+				restartable = make(map[types.UID]struct{})
+			}
+			restartable[uid] = struct{}{}
+			fallthrough
+		case orphan:
+			if p.removeTerminatedWorker(uid, status, orphan) {
+				continue
+			}
```

Comment:
nit: I think this would be a more readable written out as two separate if statements instead of switch with fallthrough, wdyt?:

```
if status.restartRequested {
    restartable[uid] = struct{}{}
}

if status.restartRequested || orphan {
    if p.removeTerminatedWorker(uid, status, orphan) {
		continue
	}
}
```

Reply:
Also, as a followup are we doing anything with the return value of `removeTerminatedWorker`? It seems that in both cases we will `continue` to the next pod in the for loop, is it correct? In that case it seems that the return value of `removeTerminatedWorker` is unused?

Reply:
removeTerminatedWorker return value skips returning the terminated worker info in `workers`.  Previously, we'd have that info hang around for the next HandlePodCleanups call (so 2 seconds where the old worker was known by UID) because we had this hypothesis that volume managers might race on teardown if a new pod showed up right away.

I don't think that's relevant or needed anymore BECAUSE

a) volume managers should be using pod worker for state anyway (one part of the possible race)
b) a static pod deleted and recreated in volume manager with same UID already needs to handle correctly tearing down unneeded volume mounts from the previous version (it's a reconciler controller to known state, you can't rely on crutches like "everything is torn down before something new is added" anyway, even if volumes aren't currently mutable)

I think this also sets us up to take away the need for the wierd restart static pod loop at the end of HandlePodCleanups because of the new `desired - actual` loop in HandlePodCleanups, but didn't want to go that far here.  RestartRequested pods move from the workers to restartable anyway.

Reply:
I see, thank you for the explanation that all makes sense. In my original reading in the github diff UI, I didn't see that the below loop was also part of the for loop :) 
 
Might be worth to leave a short comment above the continue the prupose is to ensure we don't skip adding the terminated worker in the return map. 

Re volume manager: Agreed, that makes sense. We know that Volume Manager checks explicitly with the pod worker [1], it should be safe to remove the worker right away instead of waiting fort the next loop. 

[1] https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L255-L258

Reply:
Good catch on this, I was able to a) simplify this greatly because restartablePods are exactly the same as `desired - actual`, and that highlighted i needed to change the logic.